### DISCUSSION TITLE
IHostedService calling DataPortal methods

### DISCUSSION THREAD

**Discussion by skeeler88** [MarimerLLC/csla#1831] [Questions] [answered]
I have configured Csla via Startup.cs in a ASP.NET Core 3.1 web application using Razor.

I set the context manager to be Csla.AspNetCore.ApplicationContextManager and everything works as expected throughout the website.

However when I register a [IHostedService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-3.1) (it cannot access scoped services) because I need something to run every minute I get a null reference exception when calling DataPorta.Create<TheBusinessObject>

Is there a way to call DP methods independently ? (e.g. in this case there is no IHttpContextAccessor available or HttpContext; I just want a BO command to execute).

**Comment by rockfordlhotka** [comment]
What exactly is null? Can you provide the ToString of the exception?

**Reply by skeeler88** [reply]
I cannot tell exactly without stepping into Csla v5.2.1 I believe. It may even well be that I have Csla configured wrong.

Steps to reproduce:

I've tried adding ApplicationContextManager as Transient and Singleton to the IServiceProvider. Including ActivatorUtilities.CreateInstance method within the below config.ContextManager(...) method.

1. Create & configure a IHostedService (ConfigureServices) and within initial start up file. Configure method (DataPortal is set to a Local DP)

```
app.UseCsla(config =>
{
      config.ContextManager(serviceProvider.GetService<Csla.AspNetCore.ApplicationContextManager>()); // Singleton example
});
```
2. Create a Command e.g.
```

[Serializable]
public class CommandTest
        : CommandBase<CommandTest>
{

        [Create]
        private void Create()
        {
        }
}
```

3. Start website and it calls IHostedService.StartAsync(...):

```
public Task StartAsync(CancellationToken cancellationToken)
{
DataPortal.Create<CommandTest>(); // System.NullReferenceException here
return Task.CompletedTask;
}

```
Exception received:

```
{
   "ClassName":"System.NullReferenceException",
   "Message":"Object reference not set to an instance of an object.",
   "Data":null,
   "InnerException":null,
   "HelpURL":null,
   "StackTraceString":"   at Csla.DataPortal`1.Create(Type objectType, Object criteria)\r\n   at Csla.DataPortal`1.Create(Object[] criteria)\r\n   at Csla.DataPortal.Create[T](Object[] criteria)\r\n   at Csla.DataPortal.Create[T]()\r\n   at MySpecialNamespace.NotificationsHostedService.StartAsync(CancellationToken cancellationToken) in C:\\Etc.\\NotificationsHostedService.cs:line 51",
   "RemoteStackTraceString":null,
   "RemoteStackIndex":0,
   "ExceptionMethod":null,
   "HResult":-2147467261,
   "Source":"Csla",
   "WatsonBuckets":null
}
```

I presently implemented a workaround which included several POCOs, interfaces and the new keyword...

**Reply by rockfordlhotka** [reply]
If you can get the exception object you can call ToString on it (via code, or in the Immediate window) and that'll reveal the full stack trace, subexceptions, and ultimately exactly what's happening.

**Reply by skeeler88** [reply]
In Csla.AspNetCore.ApplicationContextManager because there is no HttpContext in a IHostedService.

![image](https://user-images.githubusercontent.com/32915820/95864350-478f5100-0d33-11eb-9c6a-99ceb03c632c.png)

I think I figured a workaround in the IHostedService to fake the HttpContext:

`                this.HttpContextAccessor.HttpContext = new DefaultHttpContext();
`

**Comment by rockfordlhotka** [comment, marked as answer]
OK, so from what I can find (such as https://forums.asp.net/t/2144171.aspx?Access+HttpContext+in+HostedService) the `HttpContext` is not supposed to be available to any `IHostedService`. Apparently that's a design choice in ASP.NET.

It looks like such a service should be registered as a transient service, and any services it requires should be injected into the service: https://github.com/dotnet/extensions/issues/553

I think the issue with the data portal is that it is using an API to invoke dependency injection based on the service provider captured in Startup.cs - and that should be fine - except that the service provider instance is stored in `ApplicationContext.LocalContext`, and that (in turn) is stored in `HttpContext` - meaning it isn't available to the service, because `HttpContext` isn't available to the service.

This is something I'll need to keep in mind when implementing #1068 - and also I hope that #1068 ultimately helps address this issue.

I'm somewhat surprised that your workaround actually works, in that the service provider is captured in `Startup.cs`, and yet you are able to get at it again from your fake `HttpContext`. I wouldn't have expected that new context to have any values from the original `Startup.cs` configuration.

**Reply by skeeler88** [reply]
Further insight into the workaround you are right; we do need the service provider and that was the next exception I received. This fixed the issue (with all other needed objects injected). In particular `IServiceScopeFactory`

`            var httpContext = new DefaultHttpContext
            {
                RequestServices = this.ServiceScope.ServiceProvider,
            };
            this.HttpContextAccessor.HttpContext = httpContext;`

Now if we implement authorization/authentication on these Business Objects this approach falls over.

**Reply by rockfordlhotka** [reply]
Exactly where are you putting this code?

**Reply by skeeler88** [reply]
In IHostedService implementation

The injected services

```
        public HostedService(
            IServiceScopeFactory serviceScopeFactory,
            LinkGenerator linkGenerator,
            IEmailSender emailSender,
            IServer server,
            IHttpContextAccessor httpContextAccessor)
```

```
        public Task StartAsync(CancellationToken cancellationToken)
        {
            var httpContext = new DefaultHttpContext
            {
                RequestServices = this.ServiceScope.ServiceProvider,
            };
            this.HttpContextAccessor.HttpContext = httpContext;
            var command = DataPortal.Create<CommandTest>();
            return Task.CompletedTask;
        }
```

**Comment by dazinator** [comment]
@rockfordlhotka 

When I used CSLA a long time ago, before .net core support was supported, I manged to get this all working like so:

First, create an application context manager that doesn't rely on HttpContext:

```csharp
 /// <summary>
    /// Default context manager for the user property
    /// and local/client/global context dictionaries.
    /// </summary>
    public class ApplicationContextManager : IContextManager
    {

        private AsyncLocal<IPrincipal> _user = new AsyncLocal<IPrincipal>() { Value = new global::Csla.Security.UnauthenticatedPrincipal() };
        private static ContextDictionary _globalContext;

        private AsyncLocal<ContextDictionary> _localContext = new AsyncLocal<ContextDictionary>();
        private AsyncLocal<ContextDictionary> _clientContext = new AsyncLocal<ContextDictionary>();

        private const string _globalContextName = "Csla.GlobalContext";

        /// <summary>
        /// Returns a value indicating whether the context is valid.
        /// </summary>
        public bool IsValid
        {
            get { return true; }
        }

        /// <summary>
        /// Gets the current user principal.
        /// </summary>
        /// <returns>The current user principal</returns>
        public virtual IPrincipal GetUser()
        {
            return _user.Value;
            //   return Thread.CurrentPrincipal;

        }

        /// <summary>
        /// Sets teh current user principal.
        /// </summary>
        /// <param name="principal">User principal value</param>
        public virtual void SetUser(IPrincipal principal)
        {
            _user.Value = principal;
          //  IPrincipal current = _user.Value;
            //    Thread.CurrentPrincipal = principal;
        }

        /// <summary>
        /// Gets the local context dictionary.
        /// </summary>
        public ContextDictionary GetLocalContext()
        {
            //  LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_localContextName);
            // return (ContextDictionary)Thread.GetData(slot);
            return _localContext.Value;
        }

        /// <summary>
        /// Sets the local context dictionary.
        /// </summary>
        /// <param name="localContext">Context dictionary</param>
        public void SetLocalContext(ContextDictionary localContext)
        {
            //  LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_localContextName);
            // Thread.SetData(slot, localContext);
            _localContext.Value = localContext;
        }

        /// <summary>
        /// Gets the client context dictionary.
        /// </summary>
        public ContextDictionary GetClientContext()
        {
            //if (ApplicationContext.ExecutionLocation == ExecutionLocations.Client)
            //{
            //  return (ContextDictionary)AppDomain.CurrentDomain.GetData(_clientContextName);
            //}
            //else
            //{
            //  LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_clientContextName);
            //  return (ContextDictionary)Thread.GetData(slot);
            //}
            return _clientContext.Value;

        }

        /// <summary>
        /// Sets the client context dictionary.
        /// </summary>
        /// <param name="clientContext">Context dictionary</param>
        public void SetClientContext(ContextDictionary clientContext)
        {

            //if (ApplicationContext.ExecutionLocation == ExecutionLocations.Client)
            //{
            //  AppDomain.CurrentDomain.SetData(_clientContextName, clientContext);
            //}
            //else
            //{
            //  LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_clientContextName);
            //  Thread.SetData(slot, clientContext);
            //}
            _clientContext.Value = clientContext;
        }

        /// <summary>
        /// Gets the global context dictionary.
        /// </summary>
        public ContextDictionary GetGlobalContext()
        {
            return _globalContext;
            //LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_globalContextName);
            //return (ContextDictionary)Thread.GetData(slot);
        }

        /// <summary>
        /// Sets the global context dictionary.
        /// </summary>
        /// <param name="globalContext">Context dictionary</param>
        public void SetGlobalContext(ContextDictionary globalContext)
        {

            _globalContext = globalContext;
            //LocalDataStoreSlot slot = Thread.GetNamedDataSlot(_globalContextName);
            //Thread.SetData(slot, globalContext);

        }
    }

```

Next, also implement one that _does_ use `HttpContext` - note how this one returns `false` for `IsValid` when there is no HttpContext available so CSLA know's not to use it. (UPDATE: Pasted wrong class below, just pasted correct one!)

```csharp

 /// <summary>
    /// Application context manager that uses HttpContextAccessor whenr esolving HttpContext
    /// to store context values.
    /// </summary>
    public class HttpContextAccessorContextMananger : IContextManager
    {

        private const string _localContextName = "Csla.LocalContext";
        private const string _clientContextName = "Csla.ClientContext";
        private const string _globalContextName = "Csla.GlobalContext";

        private readonly IServiceProvider _serviceProvider;

        public HttpContextAccessorContextMananger(IServiceProvider serviceProvider)
        {
            _serviceProvider = serviceProvider;
        }


        protected virtual HttpContext GetHttpContext()
        {
            var httpContextAccessor = (IHttpContextAccessor)_serviceProvider.GetService(typeof(IHttpContextAccessor));
            if (httpContextAccessor != null)
            {
                if (httpContextAccessor.HttpContext == null)
                {
                    //   Debugger.Break();
                }
                return httpContextAccessor.HttpContext;
            }

            // Debugger.Break();
            return null;
        }


        /// <summary>
        /// Gets a value indicating whether this
        /// context manager is valid for use in
        /// the current environment.
        /// </summary>
        public bool IsValid
        {
            get
            {
                var httpContext = GetHttpContext();
                return httpContext != null;
            }
        }

        /// <summary>
        /// Gets the current principal.
        /// </summary>
        public System.Security.Principal.IPrincipal GetUser()
        {
            return GetHttpContext()?.User;
        }

        /// <summary>
        /// Sets the current principal.
        /// </summary>
        /// <param name="principal">Principal object.</param>
        public void SetUser(System.Security.Principal.IPrincipal principal)
        {
            var context = GetHttpContext();
            if (context != null)
            {
                context.User = (ClaimsPrincipal)principal;
            }
            else
            {
                //  Debugger.Break();
            }

        }

        /// <summary>
        /// Gets the local context.
        /// </summary>
        public ContextDictionary GetLocalContext()
        {
            return (ContextDictionary)GetHttpContext()?.Items[_localContextName];
        }

        /// <summary>
        /// Sets the local context.
        /// </summary>
        /// <param name="localContext">Local context.</param>
        public void SetLocalContext(ContextDictionary localContext)
        {
            var context = GetHttpContext();
            if (context != null)
            {
                context.Items[_localContextName] = localContext;
            }
            else
            {
                //  Debugger.Break();
            }
        }

        /// <summary>
        /// Gets the client context.
        /// </summary>
        public ContextDictionary GetClientContext()
        {
            return (ContextDictionary)GetHttpContext()?.Items[_clientContextName];
        }

        /// <summary>
        /// Sets the client context.
        /// </summary>
        /// <param name="clientContext">Client context.</param>
        public void SetClientContext(ContextDictionary clientContext)
        {
            var context = GetHttpContext();
            if (context != null)
            {
                context.Items[_clientContextName] = clientContext;
            }
            else
            {
                //  Debugger.Break();
            }
        }

        /// <summary>
        /// Gets the global context.
        /// </summary>
        public ContextDictionary GetGlobalContext()
        {
            return (ContextDictionary)GetHttpContext()?.Items[_globalContextName];
        }

        /// <summary>
        /// Sets the global context.
        /// </summary>
        /// <param name="globalContext">Global context.</param>
        public void SetGlobalContext(ContextDictionary globalContext)
        {
            var context = GetHttpContext();
            if (context != null)
            {
                context.Items[_globalContextName] = globalContext;
            }
            else
            {
                //  Debugger.Break();
            }
        }
    }

```

Then configure CSLA to use these two context managers:

```
  Csla.ApplicationContext.WebContextManager = options.WebContextManager ?? new HttpContextAccessorContextMananger(serviceProvider);
            Csla.ApplicationContext.ContextManager = new ApplicationContextManager();

```

Now, when CSLA is used without `HttpContext` available, it should automatically use the correct context manager and vice versa.

What about `IServiceProvider`?

I had CSLA obtain the current `IServiceProvider` using the following logic:

1. Is there an `IServiceProvider` set in `LocalContext` - use it.
2. If not, fall back to one in `GlobalContext`
3. (I then optionally used this service provider to proactively check for HttpContext (via `IHttpContextAccessor`) to see if we are running in the presence of a http request, and - if present, then "upgrade" the SP to `HttpContext.RequestServices`. This step is not really necessary if you are instead prepared to use middleware in your http pipeline that sets the HttpContext.RequestServices IServiceProvider into CSLA.LocalContext for the incoming request, so that CSLA can retrieve it during step 1 above).


The pattern is basically, on application startup, once the application level `IServiceProvider` is built , it gets stored in `GlobalContext`.

However specific workloads (like for an incoming request, or perhaps a background job) can create a scoped IServiceProvider and stick it in `LocalContext` and CSLA will then use that instead - overriding the global level one. The beauty of this is this allows things like Hangfire background jobs (not just http requests) to share the current DI scope with CSLA by setting the scoped IServiceProvider into CSLA local context at the start of the job.

```csharp
  /// <summary>
    /// Anti pattern necessary to achieve DI in some places in CSLA framework  - this was written before CSLA was using IServiceProvider or had .net core support at all..
    /// </summary>
    public static class ServiceProviderCslaExtensions
    {
      
        public static void SetAsCslaGlobalContextServiceProvider(this IServiceProvider serviceProvider)
        {
            global::Csla.ApplicationContext.GlobalContext.SetServiceProvider(serviceProvider);
        }

        public static void SetAsCslaLocalContextServiceProvider(this IServiceProvider serviceProvider)
        {
            global::Csla.ApplicationContext.LocalContext.SetServiceProvider(serviceProvider);
        }

        /// <summary>
        /// Returns the local csla service provider if available, but falls back to the global csla service provider if not. Also checks for an active HttpRequest
        /// and uses HttpContext.RequestServices if found.
        /// </summary>
        /// <returns></returns>
        public static IServiceProvider GetServiceProviderFromCslaContext()
        {
            // check local first, then fallback to global.
            IServiceProvider sp = null;
            sp = global::Csla.ApplicationContext.LocalContext.GetServiceProvider();
            if (sp == null)
            {
                sp = global::Csla.ApplicationContext.GlobalContext.GetServiceProvider();
            }

            // double check if an http request scoped service provider is available.
// note: this check shouldn't be necessary if the middleware is in play..
            var httpContextAccessor = (IHttpContextAccessor)sp.GetService(typeof(IHttpContextAccessor));
            if (httpContextAccessor != null)
            {
                if (httpContextAccessor.HttpContext != null && httpContextAccessor.HttpContext.RequestServices != null)
                {
                    sp = httpContextAccessor.HttpContext.RequestServices;
                }
            }

        
            return sp;
        }      

        public static TService Locate<TService>()
        {
            var sp = GetServiceProviderFromCslaContext();           
            var implementation = sp.GetService<TService>();
            return implementation;

        }
    }

```

and here was the middleware I used:

```csharp

  appBuilder.Use(async (context, next) =>
            {
                // On every incoming request, set request services into CSLA local-context.
                context.RequestServices.SetAsCslaLocalContextServiceProvider();
                await next.Invoke();
                // Do logging or other work that doesn't write to the Response.
            });

```

**Reply by rockfordlhotka** [reply]
@dazinator thank you for the comprehensive documentation!! Nice!

A couple thoughts based on what is now in CSLA.

> First, create an application context manager that doesn't rely on HttpContext:

We now have the [ApplicationContextManagerTls](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Core/ApplicationContextManagerTls.cs) as a type in the framework so you can just configure to use it.

> What about IServiceProvider?

This is now automatically captured on your behalf by the `AddCsla` or `UseCsla` methods in `Startup.cs`

> Then configure CSLA to use these two context managers:

I'm not 100% sure why the background task needs a web context manager at all? The point of the web context manager is to support the weird scenario in ASP.NET where some of your code relies on HttpContext, and other code does not, so both context managers are maintained. But if your code is running in an environment (like a background task) where there is no HttpContext, then the web context manager should be null.

**Reply by dazinator** [reply]
Thanks @rockfordlhotka 

> This is now automatically captured on your behalf by the AddCsla or UseCsla methods in Startup.cs

This probably results in csla capturing the "Application wide" (aka ApplicationServices) IServiceProvider which should be only used as a fallback imho when there isn't a more specific IServiceProvider available. This is the one I've suggested csla should put in GlobalContext and use if there is no other IServiceProvider overriding it in local context.

> I'm not 100% sure why the background task needs a web context manager at all? The point of the web context manager is to support the weird scenario in ASP.NET where some of your code relies on HttpContext, and other code does not, so both context managers are maintained. But if your code is running in an environment (like a background task) where there is no HttpContext, then the web context manager should be null.

The background task doesn't need a web context manager.. csla needs a context manager! The trouble is that csla needs to use the "right" context manager based on the context under which csla is being used. The web context manager is what csla should use when http context is available because context needs to travel with http context and asp.net never guaranteed a single thread would be involved in the process of a request so tls is not an appropriate backing storage for this scenario. However for a background job where there is no httpcontext, a context manager based on AsyncLocal would be more appropriate. As far as I recall (when I last checked in 2017) we still needed to register / set both context managers on application startup so csla could "pick the right one" based on which one was "valid" for use at that moment in time. This would then result in different context managers being used during a web request, and outside of a request - say in a background job.

The IServiceProvider built on startup should be put in global context imho and csla use it as a fallback if there isn't a more specific one set in local context. The idea behind setting an IServiceProvider in local context allows background jobs to create scopes and "share" those scopes with csla, it also allows http requests to do the same.. i.e asp.net core middleware could set the current (request scoped) IServiceProvider into csla local context at the start of a request.

**Reply by rockfordlhotka** [reply]
CSLA captures (or creates) a top-level context, that's true. The data portal then creates/uses a scoped context for each server-side root operation to support scoping for things like database connections, etc.

The context is maintained in `LocalContext` because it can't flow over the network to an app server (so `ClientContext` nor the soon-to-be-obsolete `GlobalContext` are applicable).

No doubt a hosted service needs its own scope. It seems to me that the initialization code for a hosted service should (must) handle some of this. Possible via an extension method. Perhaps something like

```c#
  app.UseCsla().WithHostedServiceSupport();
```

I would think that the hosted service (no HttpContext) should use the _standard_ context manager that relies on `AsyncLocal`. The reason being that `AsyncLocal` is supposed to automatically flow through various contexts, so we don't have to worry about it. Basically like HttpContext, but for non-HttpContext environments.

**Reply by rockfordlhotka** [reply]
Part of the challenge right now, today, is that we're dealing with a design that came into being in 2005 with .NET 2.

There's a proposed change for CSLA 6 that would [radically alter how ApplicationContext is managed](https://github.com/MarimerLLC/csla/issues/1068).

**Reply by dazinator** [reply]
> The data portal then creates/uses a scoped context for each server-side root operation to support scoping for things like database connections, etc.

Im not sure this the best approach.. Basically asp.net core middleware already creates a scope for a request and that scope has an IServiceProvider associated with it -  If CSLA creates its own scope, it can end up creating a new instance of services (i.e during a reuqest) when technically there is already one instance in scope. This is where telling csla to "use this IServiceProvider" instead of the globally available "application level" IServiceProvider is advantageous because:

- in a background job (IHostedService) you typically create a new DI scope which has an IServiceProvider. Currently, csla won't care about that, csla will try and create its own scope from the application level IServiceProvider. This means it won't be running within the same scope as the background job.
- during a request as I've mentioned there is already a DI scope for the request, csla just needs to use it, it will even be disposed of at the end of the request automatically - csla just needs to use the IServiceProvider associated with that scope rather than the one at application level.

If CSLA only care about using the correct ISeviceProvider, it would not need to know about di scopes. The user would create a scope, the scope has an IServiceProvider and CSLA would be told to use that rather that the application level one. This also means CSLA doesn't need to dispose of scopes. The user / application controls when scopes are created and disposed. In aspnet core that would (typically be):

1. One scope per reuqest.
2. A scope for any background job / background unit of work

> The context is maintained in LocalContext because it can't flow over the network to an app server (so ClientContext nor the soon-to-be-obsolete GlobalContext are applicable)

I didnt realise globalcontext was on its way out the door! I know about client context vs local context etc. Is the idea that global context be replaced by DI singleton or something? Either way, my proposal doesn't change really - store the application level IServiceProvider in whatever the equivalent of global context would be. The IServiceProvider for a current scope should be set by a user after creating a new di scope in their application and would be set in local context. Csla checks local context first then falls back to global context (equivalent location)

> I would think that the hosted service (no HttpContext) should use the standard context manager that relies on AsyncLocal. The reason being that AsyncLocal is supposed to automatically flow through various contexts, so we don't have to worry about it. Basically like HttpContext, but for non-HttpContext environments.

Yes I agree, that's what I use in non http context scenarios.

I was also wondering whether its safe now to use that same AsyncLocal based context manager but in Web scenarios (i.e during a request) in asp.net core too. Im not 100% sure so I am sticking with an HttpContext (IHttpContextAccesor) based one at the moment. I know state flows accross async awaits with AsyncLocal but it implies there is one top level thread from which AsyncLocal state will flow down. For an aspnet core request I am not sure if there is always one top level thread and the danger is if the asynclocal state is set in the wrong thread it won't flow to where you expect. For this reason I have stuck with using an HttpContext based context manager in aspnet core request processing scenarios!

**Reply by rockfordlhotka** [reply]
GlobalContext has always flowed from client to server _and back to the client_, and that has never actually worked in an async scenario (technically it flows back, but _you_ are left to merge it into the existing context).

I don't plan to figure out a merge algorithm to auto-merge globalcontext from the server back into the client context, and since nobody has complained about losing context over the past few years, I figure it is obviously something nobody actually uses 😀

**Reply by rockfordlhotka** [reply]
Your comment about scoped DI is interesting, not least because I have no idea how to actually _get_ the scoped provider created by ASP.NET. They intentionally make it hard/impossible to get at the provider because it is supposed to be ambient.

Do you have any idea how to access the existing provider instance?

**Reply by dazinator** [reply]
> GlobalContext has always flowed from client to server and back to the client, and that has never actually worked in an async scenario (technically it flows back, but you are left to merge it into the existing context).

Forgive me, perhaps I am confused as it has been a few years since I last worked with Csla and getting it to work under aspnet core :-) - wherever the IServiceProviders need to be stored the basic idea is:

- have a globally accessible one which is used when there is no more specific DI scope in play
- have one which flows with the context of the current caller (httpcontext, thread, task etc) which can override the global one, csla will use this if its set)
- neither of these should be serialised to the remote server!

**Reply by rockfordlhotka** [reply]
This has been the context model since CSLA 2:

* LocalContext - local to the client (per-thread or HttpContext)
* ClientContext - local to the client (per-thread or HttpContext), and flows to the server with each data portal request
* GlobalContext - (obsolete) local to the client (per-thread or HttpContext), flow to the server and back to client with each data portal request

**Reply by dazinator** [reply]
Thats odd about global context. Its possible I was "getting away with it" when placing the application level IServiceProvider in global context because we weren't using a remote server. Or could there be Csla code that doesn't serialise non serializable objects? Either way if its wrong and I was fortunate I wouldn't condone putting it in global context. However the basic mechanism / principal remains the same - put it somewhere globally accessible to csla, but that doesn't flow to the remote server - and on the remote server itself, its somewhere globally accessible that doesn't flow to a client.

**Reply by rockfordlhotka** [reply]
In terms of the DI provider (ServiceProvider), when you call `AddCsla` during app startup CSLA will [capture the IServiceCollection instance](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Configuration/ConfigurationExtensions.cs#L37). 

That is a fallback. In _most cases_ the `UseCsla` method will capture the ServiceProvider already created by .NET. For example, [in a web setting](https://github.com/MarimerLLC/csla/blob/49255247a0f8f782c6e72c757b475da89af33940/Source/Csla.Web.Mvc.Shared/WebConfiguration.cs#L33).

Only if the provider somehow wasn't captured in a `UseCsla` method will one be created using the IServiceCollection. That should only happen if you are self-hosting (console app or something) and didn't otherwise create a provider.

In any case, the top-level provider is [stored in a static](https://github.com/MarimerLLC/csla/blob/8440f9dc8f42afb8e9f3922620c32ad1b66cbcec/Source/Csla.Shared/ApplicationContext.cs#L822), because this instance is generally a per-process thing, not per-thread or per-request.

In all cases, where CSLA needs to resolve anything via DI, it uses a scoped provider for that purpose. If a scoped provider doesn't exist [one is created](https://github.com/MarimerLLC/csla/blob/8440f9dc8f42afb8e9f3922620c32ad1b66cbcec/Source/Csla.Shared/ApplicationContext.cs#L852) and maintained in `LocalContext`, because this is per-request or per-thread.

It is important to understand that [you can set the IServicesScope](https://github.com/MarimerLLC/csla/blob/0f2abbda1a1185c54617d6d937704e23acf95f52/Source/Csla.Shared/Configuration/Fluent/CslaConfiguration.cs#L163) on the current context via the CSLA fluent config model. So if one already exists you can have CSLA use that, otherwise it'll create its own.

**Reply by rockfordlhotka** [reply]
Regarding GlobalContext - you just got lucky because you didn't have a remote server 😄

**Reply by rockfordlhotka** [reply]
Also of note regarding DI, is that in CSLA 6 developers will be [required to initialize DI to use CSLA](https://github.com/MarimerLLC/csla/issues/1738).

This will resolve a host of issues, because CSLA shouldn't need to manage any service provider references itself. The goal is that CSLA can just use DI, knowing that there's already a service provider available.

The cascade effect of this change will be, I suspect, quite large. Probably comparable to where we introduced the use of generics back in 2005.

**Reply by dazinator** [reply]
> The goal is that CSLA can just use DI, knowing that there's already a service provider available.

If this mandates that the application manages the IServiceProvider and DI scopes and can just tell csla at any given time which IServiceProvider it should be using (i.e will flow but is isolated like local context does / is) then I think this is the change im waiting for! If its something different I'll still be excited to check it out ;-)

**Comment by dazinator** [comment]
> Do you have any idea how to access the existing provider instance?

Yes, but I should say that aspnet core creates this scope specifically for a request - but other user / application code is perfectly entitled to create separate DI scopes and csla won't know how to access them at all because there will be no singular way for csla to know about which scope it should be using unless something tells it. For a request, middleware can access HttpContext.RequestServices (and outside a middleware you can check IHttpContextAccessor to get current HttpContext and if there is one get RequestServices) and gives you the IServiceProvider associated with the IServiceScope that asp.net core has created for the reuqest and will dispose at the end of the request. For a background job, the user might create their own DI scope and have their own IServiceProvider which csla can't automatically know about. Csla can know about the application level (root / default scope if you like) because it gets access to this on startup, but csla can't assume that just because there is an httpcontext in play that this is the scope it should be using (this is a fairly safe assumption most of the time but not technically correct) because user code might create a new DI scope even during a reuqest, and they might want csla to run within their more granular scope than the request. Likewise in a background job a user might create several distinct scopes and run csla within each of those, or just one. If a scope is in play, a new / different IServiceProvider is in play, and csla needs to be told what that is, and this has to be local to the caller (I.e propagated similar to local context)

**Reply by rockfordlhotka** [reply]
See my concurrent reply to the previous thread - if you do create your own scope, then you obviously have that scope and can provide it to CSLA for its use.

**Comment by dazinator** [comment]
> It is important to understand that you can set the IServicesScope on the current context via the CSLA fluent config model. So if one already exists you can have CSLA use that, otherwise it'll create its own.

Thanks I just read the concurrent reply!

This is pretty much almost there then! The slight issue I think is remaining is that:

1. For a request, we can't access the requests  IServiceScope only the IServiceProvider associated with that scope so we can't prevent csla from creating a new scope when we have one in play already.
2. Imho csla shouldn't need an IServiceScope handed to it - this is owned and disposed by the application, csla just needs to operate within a given scope by using the same IServiceProvider as that scope - so csla really just needs to allow the caller to set an IServiceProvider not an IServiceScope as far as I can tell. This solves problem 1 above as csla can be given the IServiceProvider from HttpContext.RequestServices, but it also solves other scenarios in the same way, i.e in a  background job or another location (unit test?) where a user creates a new DI scope, they can hand csla the scopes IServiceProvider.

> In all cases, where CSLA needs to resolve anything via DI, it uses a scoped provider for that purpose. If a scoped provider doesn't exist one is created

Could you clarify why Csla needs to create a new DI scope? Does it dispose of this scope? If it creates a scope that is never disposed then this basically suffers from some issues. For example IDisposable transients will continue to build up until out of memory exception occurs (di scopes track IDisposable transients and they don't get released until the scope is disposed)

**Reply by dazinator** [reply]
I think your upcoming planned work you hinted at may address this.. if the intention is to it make it so that csla doesn't create a scope and the applications must tell csla what scope its operating within (by giving csla access to an IServiceProvider) then this would probably any remaining issues / nitpick I have mentioned!

**Reply by rockfordlhotka** [reply]
> For a request, we can't access the requests IServiceScope only the IServiceProvider associated with that scope so we can't prevent csla from creating a new scope when we have one in play already.

If, at the top of the request (like in a data portal interceptor) you set the IServiceScope manually, then CSLA will use your scope and won't create its own.

> Imho csla shouldn't need an IServiceScope handed to it

Oh totally!! Unfortunately right now there's no way around this, because DI is a "turtles all the way down" sort of thing. Either you use DI from top to bottom, or you need to hack it into the middle.

Having CSLA require a DI container, and rely on DI to create the top level object in a call chain (for example, the data portal server instance) is a major breaking change to all existing code. And that's what I propose for CSLA 6.

I didn't want to take such a major step right now, when so many people are just moving (or thinking about moving) from .NET Framework into this brave new world of .NET Core and .NET 5. That's a big thing by itself, without having a lot of serious changes in terms of using CSLA.

Honestly, I don't even know the full extent of the breaking changes that'll be incurred by the DI change in CSLA 6 - but I'm sure there'll be a lot of them.

**Reply by rockfordlhotka** [reply]
> by giving csla access to an IServiceProvider

See, this should never happen! If DI is used from top to bottom, then no user code (including CSLA code) should ever directly interact with the service provider. That's why the DI framework makes it really hard to get access to the service provider - because none of us should be interacting with it directly.

**Reply by rockfordlhotka** [reply]
That said, (you can probably tell I'm running through all this in my mind), CSLA _will need access to the service provider_ because we are implementing method-level injection and that's not supported by DI normally.

In other words, we MUST HAVE THE PROVIDER so we can resolve types expressed via `Inject` parameters on the data portal operation methods.

So I guess in the end, some of today's implementation won't be all that different. At the same time, a number of other things become possible if we can count on DI being available instead of using the provider pattern throughout CSLA like today.

**Reply by dazinator** [reply]
> In other words, we MUST HAVE THE PROVIDER so we can resolve types expressed via Inject parameters on the data portal operation methods

Yeah!

> That's why the DI framework makes it really hard to get access to the service provider -

Its not actually hard to access the service provider its really easy - Do you register any csla services as scoped or transient? You can have IServiceProvider injected into any of those just add it as a constructor param. Perhaps you could do this - if for example data portal was scoped or transient, it could have IServiceProvider injected which should be current scope (same scope as the service itself has been created in). All it needs to do is use that for resolving inject dependencies and bobs your uncle? However with this scheme you'd have to ensure DataPortal was being created in the correct scope - in other words, resolved using the correct IServiceProvider and this would come back to what has been discussed above- caller having a way to set IServiceProvider for csla to use in local context.

**Reply by rockfordlhotka** [reply]
Today there's no guarantee that there's a CSLA type created via DI. In CSLA 6 there will be a guarantee though, so I think you are correct about being able to get the current service provider injected. At least I hope so 🤞
